0xlmanini

high

# Amount of exiting tokens isn't checked when calling SingleSidedLiquidityVault#claimRewards()

## Summary
Missing check on amount of rewards to be sent out can lead to an overflow.

## Vulnerability Detail
When executing claimRewards(), both _claimInternalRewards() and _claimExternalRewards() lack a check on the amount of reward tokens being sent out: in the event the Vault calculates that the sender is to receive an amount of tokens larger than the Vault's balance, this can lead to an overflow which locks the user's rewards in the Vault.
The following test was added to WstethLiquidityVaultFork.t.sol to confirm the finding.
```solidity
function test_showClaimRewardsHasStrayFunds() public {
        _withdrawTimeSetUp();

        vm.startPrank(alice);

        //lower externalReward balance in liquidityVault
        liquidityVault.claimRewards();        
        liquidityVault.claimRewards();

        //trigger revert caused by overflow
        vm.expectRevert("TRANSFER_FAILED");
        liquidityVault.claimRewards();
    
        vm.stopPrank();
    }
```

## Impact
Rewards locked in Vault, stray funds

## Code Snippet
Bug may be found at
[_claimInternalRewards](https://github.com/sherlock-audit/2023-02-olympus/blob/main/src/policies/lending/abstracts/SingleSidedLiquidityVault.sol#L623-L634)
[_claimExternalRewards](https://github.com/sherlock-audit/2023-02-olympus/blob/main/src/policies/lending/abstracts/SingleSidedLiquidityVault.sol#L636-L648)

## Tool used

Manual Review

## Recommendation
Consider  calculating the amount of funds to be sent as the floor between the Vault's balance in rewardToken and the amount of reward calculated, as such:
```solidity
 uint256 vaultBal = ERC20(rewardToken.token).balanceOf(address(this));
 uint256 transferAmount = (reward - fee) < vaultBal
      ? (reward - fee)
      : vaultBal;
 if (transferAmount > 0) ERC20(rewardToken.token).safeTransfer(msg.sender, transferAmount);
```